lib/: Misc build and compiler warning fixes

andreittr opened the pull request:
### Description of changes

This changeset fixes a number of minor build errors and compiler warnings; more details in the commit descriptions, in brief:
- nolibc: switch away from syscalls for `getcwd` and `getdents64` to enable native builds
- posix-netlink: add missing select of `ukstreambuf`
- ukdebug, posix-process, ukfs-ramfs, uksparsebuf: silence compiler warnings

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

andreittr committed:
lib/nolibc: Fix native build of readdir & getcwd

This change fixes the nolibc implementation of readdir() and getcwd() to
successfully build in a minimal native configuration that lacks
syscall-shim by switching away from raw syscalls to:
- readdir: the compatible getdents64 lib wrapper
- getcwd: the internal uk_sys_getcwd function

Checkpatch-Ignore: SPDX_LICENSE_TAG
Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/posix-netlink: Config.uk select ukstreambuf

This change adds a missing select for ukstreambuf to posix-netlink's
config, as it is a hard dependency of any netlink driver.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/ukdebug: Mark crash function noreturn

This change marks the internal _uk_crash function as noreturn, silencing
a compiler warning.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/ukfs-ramfs: Error out on dir data corruption

This change has ramfs treat inconsistent directory data gracefully,
returning error instead of crashing the kernel. This also fixes several
compiler warnings.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/uksparsebuf: Define error path for separate op

This change defines an error path for calling the separate operation on
an empty sparsebuffer, instead of relying on an assert, thus preventing
an uninitialized use in the case of a bug in the implementation, thereby
also silencing a compiler warning.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
arch/x86/ectx: Convert complex cases into blocks

This change adds braces to case labels that declare variables, creating
true code blocks to correctly scope the lifetime of these variables and
prevent errors on some compiler versions.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented on lib/posix-process/exit.c:
```diff
@@ -304,12 +304,12 @@ __noreturn void exit(int status)
 	 */
 	uk_syscall_e_exit_group(status);
-	UK_BUG(); /* noreturn */
+	UK_CRASH("exit_group syscall returned"); /* noreturn */
```
This looks more like a workaround. From what I see the problem originates from the fact that we can't declare syscalls (`uk_sys_exit_group()` in this case) as `__noreturn`. So perhaps we should provide a way to declare syscalls as noreturn? CC: @mogasergiu

Also, generally speaking, as these checks are a common pattern in our codebase: If we make a call *from* a noreturn function *to* a noreturn function, is  a check really necessary? Is it possible for a noreturn function one day to accidentally return?

andreittr replied:
> If we make a call from a noreturn function to a noreturn function, is a check really necessary?

If both functions are declared noreturn no checks are needed (indeed, the compiler will just optimize them away as dead code). I believe most of these checks you mention lie in places where we _assume_ the function called to not return, without the compiler being able to know about or enforce this in any way (as we have here with the exit syscall).

michpappas replied:
Exactly my thought (question was rhetoric, as execution should be deterministic). So in this particular case a check is only meaningful if we don't provide a way to declare syscalls as noreturn. And if we do indeed enable `UK_BUG()` unconditionally, even without declaring the syscall as noreturn, `UK_BUG()` would suffice.

michpappas commented on lib/ukfs-ramfs/ramfs.c:
```diff
@@ -276,7 +276,9 @@ unsigned char ramfs_dentry_type(const struct ramfs_dentry *dp)
 		return DT_UNKNOWN;
 	default:
-		UK_BUG();
+		uk_pr_err("Invalid dentry type %d; likely bug or corruption\n",
```
This looks more like a workaround, as `UK_BUG()` should be the proper way to go in such cases. Similarly for the change below, as well as for that in `ukfs-ramfs`. IMO it should not be possible to disable `UK_BUG()`. Unlike `UK_ASSERT()` which is there to catch potential bugs, calling `UK_BUG()` implies certainty that a bug is there. To make things worse, disabling it alters the program flow, allowing execution to proceed if assertions are disabled. Shall we update libukdebug to make `UK_BUG()` always enabled similarly to `UK_CRASH()`?

andreittr replied:
> UK_BUG() should be the proper way to go in such cases. 

This update gives the ramfs code defined recoverable behavior on invalid directory contents (DT_UNKNOWN or EIO) instead of a hard system crash as before implied by `UK_BUG()`, which one of these is appropriate is not that obvious; at first I thought a hard-crash was OK, but further consideration pushed me more towards erroring out, as the latter would give at least the workload an opportunity to gracefully shutdown.

> Shall we update libukdebug to make UK_BUG() always enabled similarly to UK_CRASH()?

In that case, would there remain a difference between the two, or would they be essentially synonyms?

michpappas replied:
> This update gives the ramfs code defined recoverable behavior on invalid directory contents (DT_UNKNOWN or EIO) instead of a hard system crash as before implied by UK_BUG(), which one of these is appropriate is not that obvious; at first I thought a hard-crash was OK, but further consideration pushed me more towards erroring out, as the latter would give at least the workload an opportunity to gracefully shutdown.

Okay, but does this case justify different handling than any other case we would normally use `UK_BUG()` on?

> In that case, would there remain a difference between the two, or would they be essentially synonyms?

Functionally almost, as `UK_CRASH()` takes a paramter. In practice, I believe they have different semantics, as we would normally use `UK_CRASH()` in non-recoverable conditions of different type (panic). As is, `UK_BUG()` is pretty dangerous when assertions are disabled.

andreittr replied:
> Okay, but does this case justify different handling than any other case we would normally use UK_BUG() on?

I believe my initial use of `UK_BUG` was overly strict and wrong in this context; the right thing for a filesystem to do when encountering _data corruption_ is to error out, not crash the kernel. This fix goes beyond silencing a warning; it actually changes behavior to something I think makes more sense than before.

> use UK_CRASH() in non-recoverable conditions of different type (panic)

Do you mean that `UK_BUG()` should be used for recoverable conditions (e.g., an oops followed by a graceful shutdown), or that `UK_BUG()` is meant for non-recoverable conditions of a different type than panic? I am in favour of the former, but I think it needs more careful consideration than the scope of this PR allows.

Either way, I believe we agree that a `UK_BUG()` should never return to its calling context, as by using it you are basically saying "if we reach this line something has gone terribly wrong, do not continue". Any case where it made sense to return would be better served by a warning or such.

andreittr committed:
lib/posix-process: Declare mainlike to not return

This change has uk_posix_process_run take a mainlike function that,
appropriately to the ukthread and multiprocess requirements, is declared
noreturn in line with the expectation that it call exit() instead of
returning. The one callsite of process_run has been patched to supply
this correct wrapper function around do_main.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr commented:
Pushed 1 extra commit silencing a function cast mismatch warning.

andreittr committed:
lib/ukdebug: Make UK_BUG an unconditional halt

Previously UK_BUG() was a wrapper around UK_ASSERT(), thus allowing
execution to continue in builds with assertions disabled. This, however,
is not how it is used throughout the codebase, where it serves as an
explicit marker of an impossible code path that, if ever reached,
denotes a fundamental bug in the kernel.
This change makes UK_BUG unconditionally halt execution regardless of
configuration, and moves it to the <uk/crash.h> header.
Callers are unafected, as <uk/assert.h> includes <uk/crash.h> by
necessity.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/posix-process: Silence compiler warnings

This change consists of minor tweaks that better define behavior and
silence compiler warnings when building with assertions disabled.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr commented:
Pushed update addressing comments with the following changes:
- added 1 commit ([74e3ab9](https://github.com/unikraft/unikraft/pull/1746/commits/74e3ab99e72971fbdec2acb37fa2e1f3872c1084)) to make `UK_BUG()` an unconditional system halt, regardless of assertions being enabled
- updated the posix-process commit to remove UK_BUG -> UK_CRASH change
- updated the title & description of ramfs commit to highlight change in behavior (as that is its main purpose; warning silencing is a side-effect)

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
